map/bookmark_helpers.cpp

```diff
@@ -199,10 +199,61 @@ void ValidateKmlData(std::unique_ptr<kml::FileData> & data)
   if (!data)
     return;
 
+  // Filter points with the duplicated coordinates.
+  /// @todo(KK): This code is a temporary solution for the filtering the duplicated points in KMLs.
+  /// When the deserealizer reads the data from the KML that uses <gx:Track>
+  /// as a first step will be parsed the list of the timestamps <when> and than the list of the coordinates <gx:coord>.
+  /// So the filtering can be done only when all the data is parsed.
   for (auto & t : data->m_tracksData)
   {
     if (t.m_layers.empty())
       t.m_layers.emplace_back(kml::KmlParser::GetDefaultTrackLayer());
+
+    kml::MultiGeometry validGeometry;
+
+    auto const & geometry = t.m_geometry;
+    for (size_t lineIndex = 0; lineIndex < geometry.m_lines.size(); ++lineIndex)
+    {
+      auto const & line = geometry.m_lines[lineIndex];
+      auto const & timestamps = geometry.m_timestamps[lineIndex];
+
+      if (line.empty())
+      {
+        LOG(LWARNING, ("Empty line in track:", t.m_name[kml::kDefaultLang]));
+        continue;
+      }
+
+      kml::MultiGeometry::LineT validLine;
+      kml::MultiGeometry::TimeT validTimestamps;
```

Comment:
Are these variables needed? Can data be written into validGeometry ?

Reply:
It is easier to work with the line and timestamps rather than writing 
`auto const & prevPoint = validGeometry.m_lines.back().back();`

Reply:
I changed it to the:
```cpp
      validGeometry.m_lines.emplace_back();
      validGeometry.m_timestamps.emplace_back();

      bool const hasTimestamps = geometry.HasTimestampsFor(lineIndex);
      if (hasTimestamps && timestamps.size() != line.size())
        MYTHROW(kml::DeserializerKml::DeserializeException, ("Timestamps count", timestamps.size(), "doesn't match points count", line.size()));

      auto & validLine = validGeometry.m_lines.back();
      auto & validTimestamps = validGeometry.m_timestamps.back();

```